Skeletal Resource aws_kms_key

dromazmj opened the pull request:
* Adds functionality to be able to test basic properties of a aws kms key.

dromazmj committed:
Initial commit of skeletal resource aws_kms_key

Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
Initial commit of skeletal resource aws_kms_key

Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
Merge branch 'md/aws_kms_key' of https://github.com/dromazmj/inspec into md/aws_kms_key

Commit:
Normalizes terms accross resources (#2649)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Formatting on XML resource (#2647)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
removes OS metadata from headers

Signed-off-by: kagarmoe <kgarmoe@chef.io>

dromazmj committed:
* Adds comments to rerun travis

Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>

jquick committed:
Add 1.x and 2.x release branches (#2651)

* Add 1.x and 2.x release branches

Signed-off-by: Jared Quick <jquick@chef.io>

jquick committed:
Bump version to 2.0.0 (#2652)

Signed-off-by: Jared Quick <jquick@chef.io>

jquick committed:
Merge pull request #2655 from chef/release-2.0

Add in release-2.0 changes to master

Commit:
Bump version to 2.0.1 by Expeditor

jquick committed:
Merge pull request #2650 from chef/kg/remove-os

removes OS metadata from headers

jerryaldrichiii committed:
Add `platform` tags and remove trailing whitespace (#2654)

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Fixes obvious formatting (#2648)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

miah committed:
Add example properties from azure integration tests. (#2659)

Remove trailing whitespace.
Clean up formatting and some rubocop issues.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

Commit:
Add one comma in all docs & deletes two repeated sentences. (#2658)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

jquick committed:
Prevent resources from loading if supports check fails (#2665)

* Prevent resources from loading if supports fail.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.2 by Expeditor

juliandunn committed:
Remove duplicated encryption key test. (#2671)

Signed-off-by: Julian C. Dunn <jdunn@chef.io>

Commit:
Bump version to 2.0.3 by Expeditor

jerryaldrichiii committed:
Add `Release:` to AWS `inspec detect` example (#2672)

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Bump version to 2.0.4 by Expeditor

Commit:
Fixes merge conflict messages (#2677)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Bump version to 2.0.5 by Expeditor

chris-rock committed:
Update InSpec website (#2681)

* New styles

* Fundamental styles for home page

* Legal pages

* Community page, downloads page, tutorials page

* Docs page

* Content toggle and github badge

* Add code - highlighter

* Copy button

* Sliding content animation

* fix particle animatino on downloads page

* Scroll-to-top button

* docs sidebar links

* innocent comments on resources

* Fancy code animation

* Small update to diamond

* whatever slider

* Slider styles

* initial selection if no sliders are there

* add netifly config

Signed-off-by: Christoph Hartmann <chris@lollyrock.com>

* inspec for provisioners

* fix incorrect text on aws ssm

* fix layout javascript

post-rebase

* resource grouping in docs

* introduce groups in docs/resources

* fix minor spelling issues and move download button in new features section

Signed-off-by: Christoph Hartmann <chris@lollyrock.com>

* pass linting

Signed-off-by: Christoph Hartmann <chris@lollyrock.com>

jquick committed:
Fix bundle exec calls (#2670)

* Fix bundle exec calls and add test.

Signed-off-by: Jared Quick <jquick@chef.io>

* Add exit check for supermarket exec.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.6 by Expeditor

jquick committed:
Capture ArgumentErrors from aws. (#2673)

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.7 by Expeditor

hannah-radish committed:
Hannah Review1 (#2683)

* Smaller triangle - slowed down grid animation

* Remove slide-up from community footer - adjust slider speed

* Beginning of tutorials

* Added links

* Tutorials sliders

Commit:
Bump version to 2.0.8 by Expeditor

hannah-radish committed:
Revert download button (#2684)

Commit:
Bump version to 2.0.9 by Expeditor

jerryaldrichiii committed:
Add example profiles for AWS/Azure (#2680)

* Add example AWS profile

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

* Add example Azure profile

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Bump version to 2.0.10 by Expeditor

jquick committed:
Fix legacy reporter output to file (#2667)

* Fix legacy reporter output.

Signed-off-by: Jared Quick <jquick@chef.io>

* Wrap test in a proc to catch warnings.

Signed-off-by: Jared Quick <jquick@chef.io>

* Add output deprecation.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.11 by Expeditor

lynnvfrank committed:
Updates the year to 2018 (#2686)

The year for the generator should match the current year.

Signed-off-by: Franklin Webber <franklin@chef.io>

Commit:
Refactors the azure_resource_group doc (#2692)

* Refactors the azure_resource_group doc

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Refactoring of azure generic resources doc (#2690)

* Refactoring of azure generic resources doc

Signed-off-by: kagarmoe <kgarmoe@chef.io>

miah committed:
Add correct `supports platform` to resources. (#2674)

* Add correct `supports platform` to resources.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

* Remove 'os_family' and update platforms to specify what they did.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

* Add esx and cisco to generic resources.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.12 by Expeditor

lynnvfrank committed:
Fixes the deprecation warning text for report and output. (#2694)

* Fixes the deprecation warning text for report and output.

Remove an extra 'is being' from the 'is being is being'

Signed-off-by: Franklin Webber <franklin@chef.io>

arlimus committed:
www: fix sidebar stickiness (#2698)

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

Commit:
Bump version to 2.0.13 by Expeditor

hannah-radish committed:
HM website optimization (#2699)

* Font awesome

* Scroll to resources - done

* Overwriting swift type styles

* Fixed links

* Remove DL bin

* lint

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* lint

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* Add new breakpoints

Commit:
Bump version to 2.0.14 by Expeditor

Commit:
edits to platforms doc (#2695)

Signed-off-by: kagarmoe <kgarmoe@chef.io>

arlimus committed:
move /tutorial to /demo (#2700)

* move /tutorial to /demo

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* update netlify to build demo

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* fix gulp run path

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* ignore rubocop on www/demo

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

Commit:
Bump version to 2.0.15 by Expeditor

hannah-radish committed:
HM Mobile IE (#2705)

* Mobile media queries

* Dominik's slider broke

* slide groups fix

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

* Optimization changes

Commit:
Bump version to 2.0.16 by Expeditor

Commit:
Fixes formatting and words (#2703)

* Fixes formatting and words

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Update CHANGELOG.md to reflect the promotion of 2.0.16

Commit:
Refactors azure_virtual_machine_data_disk doc (#2701)

* Refactors azure_virtual_machine_data_disk doc

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Refactors azure_virtual_machine doc (#2702)

* Refactors azure_virtual_machine doc

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Normalizes resource (#2704)

* Normalizes resource

Signed-off-by: kagarmoe <kgarmoe@chef.io>

hannah-radish committed:
fix mobile (#2706)

* Maybe fixed mobile?

* Removed slide-right

* Change slide-right

* Bullshit commit

* add package-lock to demo

* Png fallbacks

* Small changes

* position: relative on z-index

* Overflow-x: hidden

Commit:
Spelling and punctuation check on Azure and AWS (#2657)

* Spelling and punctuation check

Signed-off-by: kagarmoe <kgarmoe@chef.io>

jquick committed:
Update shell detect to work with platforms (#2712)

* Update shell to use the same detect logic as cli detect.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.17 by Expeditor

Commit:
Update CHANGELOG.md to reflect the promotion of 2.0.17

hannah-radish committed:
Fix links on docs page (#2716)

* Fix links on docs page

Signed-off-by: hannah-radish <hmaddy@chef.io>

lynnvfrank committed:
Fixes the paths for assets loading for javascript and stylesheets (#2710)

* The paths were relative and that was causing problems on non-index pages.
* jquery apparently needs to be loaded before the menumaker.

Signed-off-by: Franklin Webber <franklin@chef.io>

hannah-radish committed:
InSpec SEO (#2725)

* Alt tags and meta descriptions

* add sitemap support

* Sitemap Priority

Signed-off-by: hannah-radish <hmaddy@chef.io>

* Jerry's feedback

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

Commit:
Bump version to 2.0.18 by Expeditor

juliandunn committed:
Now that release-2.0 has merged into master you don't need to test it anymore! (#2718)

Signed-off-by: Julian C. Dunn <jdunn@chef.io>

Commit:
Bump version to 2.0.19 by Expeditor

jquick committed:
Update maintainers file. (#2728)

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.20 by Expeditor

jerryaldrichiii committed:
package resource: Fix `brew` package detection (#2730)

* package resource: Fix `brew` package detection

This allows for package detection via `brew` to handle cases where a
particular package formula exists but is not installed.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Bump version to 2.0.21 by Expeditor

lynnvfrank committed:
Updates README to use function instead of alias (#2679)

I really liked the inclusion of the alias for the docker run version
of installing inspec. The instructions tried to make it clear that
when you create the alias that it was for only your current directory
but it sounded more like your current working directory wherever you
were so I thought that it would be better to actually make that work.

By changing it to a function you now can delay the processing of the
$(pwd) subshell so that it will mount your current working directory
when you run it. This makes it really easy to use the InSpec CLI
and the installation this way is on par with the other ways.

Signed-off-by: Franklin Webber <franklin@chef.io>

jerryaldrichiii committed:
Reword `inspec check` test's `it` block (#2721)

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Bump version to 2.0.22 by Expeditor

juliandunn committed:
Always make sure the copyright footer has the correct year (#2707)

* Always make sure the copyright footer has the correct year

Signed-off-by: Julian C. Dunn <jdunn@chef.io>

jerryaldrichiii committed:
docs: Enhance `http` resource documentation (#2715)

* Simplify `http` resource example

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

juliandunn committed:
Fix infrastructure code example and typo in parse_config_file reference (#2708)

* Fix infrastructure code example and typo in parse_config_file reference.

Signed-off-by: Julian C. Dunn <jdunn@chef.io>

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+title: About the aws_kms_key Resource
+---
+
+# aws_kms_key
```
Escape underscores in unquoted strings

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+---
```
Doc files should now end in .md.erb

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+AWS Key Management Service (AWS KMS) is a managed service that makes it easy for you to create and control the encryption keys used to encrypt your data. AWS KMS lets you create master keys that can never be exported from the service and which can be used to encrypt and decrypt data based on policies you define.
+
+Each AWS KMS Key is uniquely identified by its key-id or key-arn.
+
```
Since we have `aws_kms_keys`, you might add a reference here - "Use `aws_kms_key` to verify the properties of a single key.  Use `aws_kms_keys` to verify the properties of all or a group of keys."

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+## Syntax
+
+An aws_kms_key resource block identifies a key by key_arn or just the key.
```
Instead of "or just the key", try "or the key id"

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+      it { should be_enabled }
+    end
+
+### Test that the specified key is rotation enabled
```
Rename this matcher to be have_rotation_enabled

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+The globally unique identifier for the key.
+
+    describe aws_kms_key('arn:aws:kms:us-east-1::key/4321dcba-21io-23de-85he-ab0987654321') do
+      its('key_id') { should cmp "id" }
```
Use a realistic comparison value for the key ID

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+Specifies the date and time when the key was created.
+
+    describe aws_kms_key('arn:aws:kms:us-east-1::key/4321dcba-21io-23de-85he-ab0987654321') do
+      its('creation_date') { should be < Time.now - 10 * 86400 }
```
A comment here explaining the math would be helpful to non-technical readers.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+      its('creation_date') { should be < Time.now - 10 * 86400 }
+    end
+
+### created_days_ago
```
:100:

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+Specifies the number of days since the key was created.
+
+    describe aws_kms_key('arn:aws:kms:us-east-1::key/4321dcba-21io-23de-85he-ab0987654321') do
+      its('created_days_ago') { should be > 10 }
```
It might make more sense as an example to check that a key is *less* than 10 days old

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+### key_usage
+
+Specifies the cryptographic operations for which you can use the key. Currently the only allowed value is ENCRYPT_DECRYPT , which means you can use the key for the encrypt and decrypt operations.
```
If this only has one possible value, I'm not sure we need to expose it, or at least not document it.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+### key_state
+
+Specifies the state of the key one of "Enabled", "Disabled", "PendingDeletion", "PendingImport".
```
Might add a reference to the `be_enabled` matcher.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+### deletion_date
+
+Specifies the date and time after which AWS KMS deletes the key. This value is present only when KeyState is PendingDeletion , otherwise this value is omitted.
```
What does omitted mean?  Nil?

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+### valid_to
+
+Provides the data and time until the key is not valid.  Once the key is not valid, AWS KMS deletes the key and it becomes unusable.  This value will be null unless the keys Origin is EXTERNAL and its ExpirationModel is KEY_MATERIAL_EXPIRES.
```
This looks like AWS docs.  What is Null in Ruby-land?  Also, I don't think we're exposing ExpirationModel, so that may be confusing.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+### be_enabled
+
+The test will pass if the specified key is enabled.
```
I'd perhaps phrase this as "the key's state is `Enabled`." to be clear you're referring to KeyState.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+    end
+    
+### be_external
+Provides whether the source of the key's key material is external or from not.  If it is not external than it was created by AWS KMS.  When it is external, the key material was imported from an existing key management infrastructure or the key lacks key material.
```
I think this needs some reworking for clarity.  "from not" is not meaningful, as far as I can parse.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+
+AWS Key Management Service (AWS KMS) is a managed service that makes it easy for you to create and control the encryption keys used to encrypt your data. AWS KMS lets you create master keys that can never be exported from the service and which can be used to encrypt and decrypt data based on policies you define.
+
+Each AWS KMS Key is uniquely identified by its key-id or key-arn.
```
Use the syntax used in the resource: `key_id` instead of `key-id`, and `arn` instead of `key-arn`.

clintoncwolfe commented on lib/resources/aws/aws_kms_key.rb:
```diff
@@ -0,0 +1,93 @@
+    backend = BackendFactory.create(inspec_runner)
+
+    query = { key_id: @key_id }
+    resp = backend.describe_key(query)
```
Wrap in a `catch_aws_errors` wrapper

clintoncwolfe commented on lib/resources/aws/aws_kms_key.rb:
```diff
@@ -0,0 +1,93 @@
+    @has_key_expiration = @key[:expiration_model] == 'KEY_MATERIAL_EXPIRES'
+    @has_aws_key_manager = @key[:key_manager] == 'AWS'
+
+    resp = backend.get_key_rotation_status(query)
```
Wrap in `catch_aws_errors`

clintoncwolfe commented on test/aws/default/build/kms.tf:
```diff
@@ -0,0 +1,42 @@
+#                                 SNS Topic
+#===========================================================================#
+# Test fixture: 
+# sns_topic_1 has one confirmed subscription
```
copypasta

clintoncwolfe commented on test/aws/default/verify/controls/aws_kms_key.rb:
```diff
@@ -0,0 +1,51 @@
+  describe aws_kms_key(fixtures['kms_key_recall_hit_arn']) do
+    it { should exist}
+  end
+  describe aws_kms_key(key_id: fixtures['kms_key_recall_hit_arn']) do
```
The API doesn't care if you send an ARN or an ID, so I think this works, but you should probably be sending the key_id here rather than the ARN.

clintoncwolfe commented on docs/resources/aws_kms_key.md:
```diff
@@ -0,0 +1,175 @@
+      its { should be_external }
+    end
+    
+### have_aws_key_manager
```
I think we might want to rename this be_managed_by_aws (managed_by_aws?) for readability.

dromazmj replied:
Good idea!

clintoncwolfe reviewed (changes requested):
Very nice, thorough PR!  A few docs changes, exception wrapping, and some suggested matcher/property renaming - but very solid.  You've really hit your stride on these!

dromazmj committed:
Merge branch 'master' of https://github.com/chef/inspec into md/aws_kms_key

jquick committed:
Fix inspec check to work with platforms (#2737)

* Fix inspec check to work with platforms.

Signed-off-by: Jared Quick <jquick@chef.io>

Commit:
Bump version to 2.0.23 by Expeditor

arlimus committed:
emoji in issue template (#2743)

* emoji in issue template

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

jerryaldrichiii committed:
Move AWS/Azure tests to integration directory (#2675)

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

Commit:
Bump version to 2.0.24 by Expeditor

lynnvfrank committed:
Fixes partially indented blocks and unescaped underscores (#2731)

Moved 2 space examples 2 more spaces in. Don't be shy, show the world your code the way it was meant to be seen.

Underscores in markdown must be escaped otherwise the world goes crooked.

Signed-off-by: Franklin Webber <franklin@chef.io>

dromazmj committed:
* Clarifies some parts of the doc.
* Changes matcher have_aws_key_manager to manged_by_aws
* Fixes copypasta

Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>

dromazmj committed:
Merge branch 'master' of https://github.com/chef/inspec into md/aws_kms_key
